File: Esp32/ESP32-doit_devkit_v1/PHP/simular-esp.py
#!/usr/bin/env python3
"""
Simulador ESP — Temp Segura (Python 3, sin instalar nada / solo librería estándar)

Calcula el HMAC y manda una lectura firmada al servidor, igual que el firmware.

USO:
  python simular-esp.py            -> te pregunta la temperatura y la manda
  python simular-esp.py 12.3       -> manda temp 12.3
  python simular-esp.py 25 nosend  -> solo muestra hash + body + curl (NO envía)

Editá SECRET con el valor COMPLETO (64 hex) del dispositivo en la DB.
"""

import hashlib
import hmac
import json
import os
import sys
import time
import urllib.request
import urllib.error

# ===== CONFIG =====
BASE   = os.environ.get("ESP_BASE")   or "https://tempsegura.orbitar.dev/api"
MAC    = os.environ.get("ESP_MAC")    or "4C:11:AE:70:26:70"
SECRET = os.environ.get("ESP_SECRET") or "b45f063b9605742450c78649b78e18aca47e9b4fc23571e60e046fa3853f0462"
# ==================


def main():
    # Temperatura: por argumento o preguntando
    if len(sys.argv) > 1:
        temp = float(sys.argv[1])
    else:
        temp = float(input("Temperatura a enviar (°C): ").strip().replace(",", "."))
    nosend = len(sys.argv) > 2 and sys.argv[2] == "nosend"

    # PHP serializa los float enteros sin ".0" (8.0 -> "8"); Python los deja "8.0".
    # Para que la firma coincida, mandamos los enteros como int.
    if temp.is_integer():
        temp = int(temp)

    ts = int(time.time())

    # Cuerpo que se firma: TODO menos mac, timestamp y signature.
    data = {
        "packet_id": "pkt-%d" % ts,
        "seq": 1,
        "data": [{"temp": temp, "time": ts}],
        "local_alerts": [],
        "optional": {"firmware_version": "1.0.0", "rssi": -60},
    }

    # Canonicalización idéntica al backend: claves ordenadas (recursivo), sin espacios, sin escapar.
    json_data = json.dumps(data, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    message = (MAC + str(ts) + json_data).encode("utf-8")
    signature = hmac.new(SECRET.encode("utf-8"), message, hashlib.sha256).hexdigest()

    body = {"mac": MAC, "timestamp": ts, "signature": signature}
    body.update(data)
    body_str = json.dumps(body, ensure_ascii=False)

    print("timestamp : %d" % ts)
    print("signature : %s" % signature)
    print("body      : %s\n" % body_str)

    if nosend:
        print("curl listo:")
        print("curl -X POST \"%s/esp/sync\" -H \"Content-Type: application/json\" -d '%s'" % (BASE, body_str))
        return

    req = urllib.request.Request(
        BASE + "/esp/sync",
        data=body_str.encode("utf-8"),
        headers={"Content-Type": "application/json", "Accept": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=25) as r:
            print("===> HTTP %d" % r.status)
            print(r.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        print("===> HTTP %d" % e.code)
        print(e.read().decode("utf-8"))
    except Exception as e:
        print("Error de conexión: %s" % e)


if __name__ == "__main__":
    main()

File: Esp32/ESP32-doit_devkit_v1/PHP/simular-esp.php
<?php
/**
 * Simulador ESP por línea de comandos — Temp Segura
 *
 * Calcula el HMAC y manda una lectura firmada al servidor, igual que el firmware.
 *
 * USO:
 *   php simular-esp.php            -> manda temp 8.5
 *   php simular-esp.php 12.3       -> manda temp 12.3
 *   php simular-esp.php 25 nosend  -> solo muestra hash + body (no envía)
 *
 * Editá SHARED_SECRET con el valor COMPLETO (64 hex) del dispositivo en la DB.
 * (También se puede sobreescribir por variables de entorno ESP_BASE/ESP_MAC/ESP_SECRET.)
 */

// ===== CONFIG =====
$BASE   = getenv('ESP_BASE')   ?: 'https://tempsegura.orbitar.dev/api';
$MAC    = getenv('ESP_MAC')    ?: '4C:11:AE:70:26:70';
$SECRET = getenv('ESP_SECRET') ?: 'b45f063b9605742450c78649b78e18aca47e9b4fc23571e60e046fa3853f0462';
// ==================

$temp   = isset($argv[1]) ? (float)$argv[1] : 8.5;
$noSend = isset($argv[2]) && $argv[2] === 'nosend';
$ts     = time();

if ($SECRET === 'PEGAR_SECRET_COMPLETO_64_HEX') {
    fwrite(STDERR, "⚠️  Falta el shared_secret. Editá la línea SHARED_SECRET del script con el valor completo del dispositivo.\n");
    exit(1);
}

// Canonicalización idéntica al backend: claves de objetos ordenadas (recursivo); arrays en orden.
function canon($v) {
    if (is_array($v)) {
        $isList = array_keys($v) === range(0, count($v) - 1);
        if ($isList) return array_map('canon', $v);
        ksort($v, SORT_STRING);
        foreach ($v as $k => $x) $v[$k] = canon($x);
        return $v;
    }
    return $v;
}

$data = [
    'packet_id'    => 'pkt-' . $ts,
    'seq'          => 1,
    'data'         => [['temp' => $temp, 'time' => $ts]],
    'local_alerts' => [],
    'optional'     => ['firmware_version' => '1.0.0', 'rssi' => -60],
];

$jsonData  = json_encode(canon($data), JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
$signature = hash_hmac('sha256', $MAC . $ts . $jsonData, $SECRET);
$body      = json_encode(array_merge(
    ['mac' => $MAC, 'timestamp' => $ts, 'signature' => $signature],
    $data
), JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);

echo "timestamp : $ts\n";
echo "signature : $signature\n";
echo "body      : $body\n\n";

if ($noSend) {
    echo "curl listo:\n";
    echo "curl -X POST \"$BASE/esp/sync\" -H \"Content-Type: application/json\" -d '$body'\n";
    exit(0);
}

$ch = curl_init("$BASE/esp/sync");
curl_setopt_array($ch, [
    CURLOPT_POST           => 1,
    CURLOPT_POSTFIELDS     => $body,
    CURLOPT_HTTPHEADER     => ['Content-Type: application/json'],
    CURLOPT_RETURNTRANSFER => 1,
    CURLOPT_TIMEOUT        => 25,
]);
$resp = curl_exec($ch);
$code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
echo "===> HTTP $code\n$resp\n";
